trunner/harnesses/micropython.py

```diff
@@ -0,0 +1,221 @@
+        if MicropythonStandardHarness.is_special_test(test_path):
+            exp_result = Tools.create_regex(exp_result)
+            if re.match(exp_result, test_result):
+                test_passed = True
+        elif exp_result == test_result:
+            test_passed = True
+
+        if test_passed:
+            tests_results.append(TestResult(test_path, TestResult.PASS))
+        else:
+            tests_results.append(TestResult(test_path, TestResult.FAIL, error_msg))
+
+        return tests_results
+
+
+class MicropythonReplHarness:
+    """Class providing harness for testing micropython REPL"""
+
+    UPYTH_PROMPT = ">>>"
+    MICROPYTHON = "/bin/micropython"
+
+    def send_micropython_cmd(cmd, proc):
+        proc.sendline(cmd)
+        Tools.expect(MicropythonReplHarness.UPYTH_PROMPT, proc)
+
+        return proc.before + MicropythonReplHarness.UPYTH_PROMPT
+
+    @staticmethod
+    def harness(p):
+
+        tests_results = []
+
+        # At first echo prints path to the test
+        try:
+            Tools.expect(Tools.PROMPT, p)
+        except AssertionError as e:
+            tests_results.append(TestResult("Unknown REPL test", TestResult.FAIL, str(e)))
+            return tests_results
+
+        test_path = p.before
+
+        # sending cat with path to the test to get lines to insert in micropython terminal
+        cmd = "cat " + test_path
+        p.sendline(cmd)
+        try:
+            Tools.expect(Tools.PROMPT, p)
+        except AssertionError as e:
+            tests_results.append(TestResult(test_path, TestResult.FAIL, str(e)))
+            return tests_results
+
+        script = p.before
+        script = script.split("\n")
+
+        # sending cat with path to the expected output from test
+        cmd += ".exp"
+        p.sendline(Tools.PROMPT)
+        try:
+            Tools.expect(Tools.PROMPT, p)
+        except AssertionError as e:
+            tests_results.append(TestResult(test_path, TestResult.FAIL, str(e)))
+            return tests_results
+
+        expected_output = p.before
+
+        test_result = []
+
+        try:
+            test_result += MicropythonReplHarness.send_micropython_cmd(MicropythonReplHarness.MICROPYTHON, p)
+
+            for line in script:
+                test_result += MicropythonReplHarness.send_micropython_cmd(line, p)
+        except AssertionError as e:
+            tests_results.append(TestResult(test_path, TestResult.FAIL, str(e)))
+            return tests_results
+
+        test_result = "".join(test_result)
+
+        error_msg = f"Incorrect output\n\nExpected result:\n{expected_output}\nTest result:\n{test_result}"
+
+        expected_output = Tools.create_regex(expected_output)
+
+        if re.match(expected_output, test_result):
+            tests_results.append(TestResult(test_path, TestResult.PASS))
+        else:
+            tests_results.append(TestResult(test_path, TestResult.FAIL, error_msg))
+
+        return tests_results
+
+
+class Tools:
```

Comment:
I think that this construct of `Tools` class works, but should be changed. You use methods of that class in harnesses, but they are not defined as static, and no instance of this object is created. The pythonc way would be to either:
 - make this tools class into methods and store them in `tools.py`, or `tools/upythTools.py` module
 - move this class to the beginning of this file and:
   - make an instance of that object in each harness
   - make those functions static

Reply:
I am going to move this class to the beginning and make those functions static